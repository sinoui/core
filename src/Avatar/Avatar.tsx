import React from 'react';
import styled from 'styled-components';
import classNames from 'classnames';
import colorCss from '@sinoui/core/utils/colorCss';

export interface AvatarProps {
  /**
   * 子元素
   */
  children?: React.ReactNode;
  /**
   * 是否为密集模式
   */
  dense?: boolean;
  /**
   * 头像背景色
   */
  color?: string;
  /**
   * 添加自定义类名
   */
  className?: string;
}

/**
 * 头像组件
 */
const Avatar = styled.div.attrs(({ color = 'actionDisabled', className }) => ({
  color,
  className: classNames('sinoui-avatar', className),
}))<AvatarProps>`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: ${(props) => (props.dense ? '32px' : '40px')};
  height: ${(props) => (props.dense ? '32px' : '40px')};
  border-radius: 50%;
  font-size: ${(props) => props.theme.typography.h6};
  ${colorCss('background-color')};
  color: #fff;
  overflow: hidden;
  user-select: none;

  > img {
    width: 100%;
    height: 100%;
    color: transparent;
    object-fit: cover;
  }
`;

export default Avatar;
