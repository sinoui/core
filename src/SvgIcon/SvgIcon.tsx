import React from 'react';
import styled from 'styled-components';
import classNames from 'classnames';
import colorCss from '@sinoui/core/utils/colorCss';
import { removeUndefinedProperties } from '../utils/objects';

export interface SvgIconProps {
  /**
   * 指定图标内容
   */
  children?: React.ReactNode;
  /**
   * 自定义类名
   */
  className?: string;
  /**
   * 指定自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 设置 svg 元素的 viewBox，默认为0 0 24 24
   */
  viewBox?: string;
  /**
   * 设置图标大小。数字或者字符串。如果是数字，则单位是像素
   */
  size?: number | string;
  /**
   * 设置颜色
   */
  color?: string;
  /**
   * 指定图标的标题
   */
  title?: string;
  /**
   * 不设置图标大小。一般用于`<SvgIcon as={CustomIcon} disabledViewBox />`，这种情况下，CustomIcon已有了`viewBox`。
   */
  disabledViewBox?: boolean;
}

/**
 * SvgIcon组件
 */
const SvgIcon = styled.svg
  .withConfig({
    shouldForwardProp: (prop) => !['disabledViewBox'].includes(prop),
  })
  .attrs(
    ({
      children,
      className,
      disabledViewBox,
      viewBox = !disabledViewBox ? '0 0 24 24' : undefined,
      size = 24,
      title,
    }: SvgIconProps) =>
      removeUndefinedProperties({
        children: (
          <>
            {!!title && <title>{title}</title>}
            {children}
          </>
        ),
        className: classNames('sinoui-svg-icon', className),
        viewBox,
        size,
        'aria-hidden': title ? undefined : 'true',
        role: title ? 'img' : 'presentation',
        focusable: 'false',
      }),
  )<SvgIconProps>`
  font-size: ${(props) =>
    typeof props.size === 'string' ? props.size : `${props.size}px`};
  width: 1em;
  height: 1em;
  fill: currentColor;
  flex-shrink: 0;
  transition: ${(props) =>
    props.theme.transitions.create('color', {
      duration: props.theme.transitions.duration.shorter,
    })};
  ${colorCss('color')};
  user-select: none;
  display: inline-block;
`;

export default SvgIcon;
