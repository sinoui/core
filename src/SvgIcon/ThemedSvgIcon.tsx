import styled from 'styled-components';
import { getColorFromTheme } from '../utils/colors';
import { ColorProp } from '../types';
import SvgIcon from './SvgIcon';

export interface Props {
  color?: ColorProp;
}

const ThemedSvgIcon = styled(SvgIcon) <Props>`
  ${(props) =>
    props.color !== 'none' && `color: ${getColorFromTheme(props, 'primary')}`};
`;

ThemedSvgIcon.sinouiName = 'SvgIcon';
ThemedSvgIcon.displayName = 'SvgIcon';

export default ThemedSvgIcon;
