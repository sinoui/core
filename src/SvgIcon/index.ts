import SvgIcon, { SvgIconProps } from './SvgIcon';

export type { SvgIconProps };
export default SvgIcon;
