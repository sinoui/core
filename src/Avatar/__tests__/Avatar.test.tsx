import React from 'react';
import renderer from 'react-test-renderer';
import { MdAdd } from 'react-icons/md';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import Avatar from '@sinoui/core/Avatar';
import SvgIcon from '@sinoui/core/SvgIcon';
import avatarf from './avatarf.jpg';
import avatars from './avatars.jpg';
import avatart from './avatart.jpg';

/**
 * Avatar头像组件
 */

describe('Avatar组件 快照测试', () => {
  it('图片头像', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Avatar>
            <img src={avatarf} alt="" />
          </Avatar>
          <Avatar>
            <img src={avatars} alt="" />
          </Avatar>
          <Avatar>
            <img src={avatart} alt="" />
          </Avatar>
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('字母头像', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Avatar>H</Avatar>
          <Avatar>N</Avatar>
          <Avatar>LX</Avatar>
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('图标头像', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Avatar>
            <SvgIcon>
              <MdAdd />
            </SvgIcon>
          </Avatar>
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('是否为密集模式', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Avatar dense>
            <img src={avatarf} alt="" />
          </Avatar>
          <p />
          <Avatar dense>
            <SvgIcon as={MdAdd} />
          </Avatar>
          <p />
          <Avatar dense>H</Avatar>
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('设置不同背景颜色', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Avatar color="primary">LX</Avatar>
          <Avatar color="secondary">LX</Avatar>
          <Avatar color="warning">LX</Avatar>
          <Avatar color="error">LX</Avatar>
          <Avatar color="success">LX</Avatar>
          <Avatar color="info">LX</Avatar>
          <Avatar color="actionActive">LX</Avatar>
          <Avatar color="actionDisabled">LX</Avatar>
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
