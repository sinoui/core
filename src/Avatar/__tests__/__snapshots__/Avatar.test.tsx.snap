// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar组件 快照测试 图标头像 1`] = `
<div
  className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
  color="actionDisabled"
>
  <svg
    aria-hidden="true"
    className="SvgIcon-sc-1015m7s-0 kTyazq sinoui-svg-icon"
    focusable="false"
    role="presentation"
    size={24}
    viewBox="0 0 24 24"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      />
    </svg>
  </svg>
</div>
`;

exports[`Avatar组件 快照测试 图片头像 1`] = `
Array [
  <div
    className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
    color="actionDisabled"
  >
    <img
      alt=""
      src="avatarf.jpg"
    />
  </div>,
  <div
    className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
    color="actionDisabled"
  >
    <img
      alt=""
      src="avatars.jpg"
    />
  </div>,
  <div
    className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
    color="actionDisabled"
  >
    <img
      alt=""
      src="avatart.jpg"
    />
  </div>,
]
`;

exports[`Avatar组件 快照测试 字母头像 1`] = `
Array [
  <div
    className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
    color="actionDisabled"
  >
    H
  </div>,
  <div
    className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
    color="actionDisabled"
  >
    N
  </div>,
  <div
    className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
    color="actionDisabled"
  >
    LX
  </div>,
]
`;

exports[`Avatar组件 快照测试 是否为密集模式 1`] = `
Array [
  <div
    className="Avatar-gvflnr-0 jElJCw sinoui-avatar"
    color="actionDisabled"
  >
    <img
      alt=""
      src="avatarf.jpg"
    />
  </div>,
  <p />,
  <div
    className="Avatar-gvflnr-0 jElJCw sinoui-avatar"
    color="actionDisabled"
  >
    <svg
      aria-hidden="true"
      className="SvgIcon-sc-1015m7s-0 kTyazq sinoui-svg-icon"
      fill="currentColor"
      focusable="false"
      height={24}
      role="presentation"
      size={24}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      />
    </svg>
  </div>,
  <p />,
  <div
    className="Avatar-gvflnr-0 jElJCw sinoui-avatar"
    color="actionDisabled"
  >
    H
  </div>,
]
`;

exports[`Avatar组件 快照测试 设置不同背景颜色 1`] = `
Array [
  <div
    className="Avatar-gvflnr-0 ipytFM sinoui-avatar"
    color="primary"
  >
    LX
  </div>,
  <div
    className="Avatar-gvflnr-0 jLOYMD sinoui-avatar"
    color="secondary"
  >
    LX
  </div>,
  <div
    className="Avatar-gvflnr-0 pWIdP sinoui-avatar"
    color="warning"
  >
    LX
  </div>,
  <div
    className="Avatar-gvflnr-0 bocrdO sinoui-avatar"
    color="error"
  >
    LX
  </div>,
  <div
    className="Avatar-gvflnr-0 jnqvZL sinoui-avatar"
    color="success"
  >
    LX
  </div>,
  <div
    className="Avatar-gvflnr-0 jtimoL sinoui-avatar"
    color="info"
  >
    LX
  </div>,
  <div
    className="Avatar-gvflnr-0 cra-DDW sinoui-avatar"
    color="actionActive"
  >
    LX
  </div>,
  <div
    className="Avatar-gvflnr-0 cSfgjc sinoui-avatar"
    color="actionDisabled"
  >
    LX
  </div>,
]
`;
